flink-core/src/test/java/org/apache/flink/core/fs/SafeFileSystemFactoryTest.java

```diff
@@ -0,0 +1,183 @@
+        testProtectionDomainLeak(
+                new TestFileSystemFactory() {
+
+                    @Override
+                    public FileSystem create(URI fsUri) {
+                        boolean leak = false;
+                        for (ProtectionDomain protectionDomain : getProtectionDomains()) {
+                            if (protectionDomain.getClassLoader()
+                                    instanceof ChildFirstClassLoader) {
+                                leak = true;
+                                break;
+                            }
+                        }
+                        Assertions.assertTrue(leak);
+                        return null;
+                    }
+                });
+    }
+
+    @Test
+    void testProtectionDomainDoesNotLeakWithWrappedFileSystemFactory() throws Exception {
+        testProtectionDomainLeak(
+                SafeFileSystemFactory.of(
+                        new TestFileSystemFactory() {
+
+                            @Override
+                            public FileSystem create(URI fsUri) {
+                                for (ProtectionDomain protectionDomain : getProtectionDomains()) {
+                                    Assertions.assertFalse(
+                                            protectionDomain.getClassLoader()
+                                                    instanceof ChildFirstClassLoader,
+                                            "ClassLoader leaked.");
+                                }
+                                return null;
+                            }
+                        }));
+    }
+
+    private void testProtectionDomainLeak(FileSystemFactory factory) throws Exception {
+        final URL testClassLocation =
+                getClass().getProtectionDomain().getCodeSource().getLocation();
+        final ChildFirstClassLoader leakyClassLoader =
+                new ChildFirstClassLoader(
+                        new URL[] {testClassLocation},
+                        Thread.currentThread().getContextClassLoader(),
+                        new String[] {"java."},
+                        error -> {
+                            // No-op.
+                        });
+        try (TemporaryClassLoaderContext ignored =
+                TemporaryClassLoaderContext.of(leakyClassLoader)) {
+            final Class<?> clazz =
+                    Class.forName(ProtectionDomainLeakTest.class.getName(), true, leakyClassLoader);
+            final Constructor<?> declaredConstructor =
+                    clazz.getDeclaredConstructor(FileSystemFactory.class);
+            declaredConstructor.setAccessible(true);
+            declaredConstructor.newInstance(factory);
+        }
+        Assertions.assertEquals(1, getLoadedFlinkClasses(leakyClassLoader).size());
+        Assertions.assertEquals(
+                ProtectionDomainLeakTest.class.getName(),
+                getLoadedFlinkClasses(leakyClassLoader).get(0).getName());
```

Comment:
So the classloader is leaky in any ways and we really just try to contain the leak to CL and not also PD?

Reply:
yes, it's all about leaking the reference to the classloader, which makes it ineligible for the GC once the job finishes.

Reply:
Out of curiosity: CL means class loading, doesn't it? What about PD? 🤔

Reply:
PD=ProtectionDomain